### PR TITLE
[WIP] Update Pubspec.yaml 

### LINKED ISSUE #16: action not updating pubspec.yaml file (it updates pubspec.lock only instead)
## The Issue
At its current state the Bot is does not work as advertised.
The file updated is `pubspec.lock` but when the PR is accepted Flutter runs `flutter packages get` (with info from the yaml file) and overrides the changes that the bot does.

## Ideal Situation
Make changes to the `pusbspec.yaml` file and commit them as a part of the PR.

### Reading into how this do be implemented.

1. Somehow convert data from `flutter pub outdated` and convert that into a suitable form.
image
2. Take these dependencies can be automatically updated (following Semver) and patch the pubspec.yaml file.
[Yaml for JS](https://www.npmjs.com/package/yaml)

## References:
* https://github.com/flutter/flutter/issues/13428
* https://github.com/flutter/flutter/issues/40657
* https://github.com/flutter/flutter/issues/29301
* https://gist.github.com/spidgorny/ed475cbbe303e1c09c0f6c9f9f57dcad

### PR DESCRIPTION
**Pull request by preetjdp**
# What this does
A attempt to manually update the pubspec.yaml

Fixes https://github.com/tianhaoz95/update-flutter-packages/issues/16

Loops through each dependency in the project, fetches the versions of the dependency from the api, checks for a patchable version and patches it.

# Quirks 
Still need to vet for edge cases such as
* Self Hosted Package.
* Package from git
* Add more.

@ilteoood Does this seem like a viable direction to go with?

### CODE REVIEW

File: src/utils/asyncForEach.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export async function asyncForEach<T>(array: T[], callback: (item: T, index: number, allItems: T[]) => void) {
```
Comment (ilteoood) [inline src/utils/asyncForEach.ts:1]: I don't like this asyncForEach.
You are using the async/await, you should avoid the callbacks

File: src/upgrade.ts [hunk 2]
```diff
@@ -1,13 +1,69 @@
...
+  let pubspecFilePath = `${context.flutterProjectWorkspace}/pubspec.yaml`
+  try {
+    const pubspecFile = readFileSync(pubspecFilePath, 'utf8')
+    let docYaml = parseDocument(pubspecFile)
+    let dependencyMap: YAMLMap = docYaml.get('dependencies')
+    if (!dependencyMap) { throw "No Dependecies Found" }
+
+    let dependeciesList = dependencyMap.items
+    if (dependeciesList.length == 0) { throw "No Dependecies Present" }
+    await asyncForEach(dependeciesList, async (dependency) => {
```
Comment (ilteoood) [inline src/upgrade.ts:29]: Instead of this I suggest you to:

File: src/upgrade.ts [hunk 3]
```diff
@@ -1,13 +1,69 @@
...
+      const dependencyKey: Scalar = dependency.key
+      const dependencyValue: Scalar = dependency.value
+
+      let dependencyName = dependencyKey.value
+      let dependencyVersion = dependencyValue.value?.replace('^', '') ?? undefined
+
+      if (!dependencyName) { return }
+      if (!dependencyVersion) { return }
+
+      if (['flutter'].includes(dependencyName)) {
```
Comment (ilteoood) [inline src/upgrade.ts:39]: 1) apply a filter, in order to remove this

File: src/upgrade.ts [hunk 4]
```diff
@@ -1,13 +1,69 @@
...
+        return
+      }
+
+      let latestVersion = await getLatestPackage(dependency)
```
Comment (ilteoood) [inline src/upgrade.ts:43]: 2. use a map in order accumulate an array of promises

File: src/upgrade.ts [hunk 5]
```diff
@@ -1,13 +1,69 @@
...
+      if(!latestVersion) {
```
Comment (ilteoood) [inline src/upgrade.ts:44]: 3. Use Promise.all to "convert" the array of promises into an array of string

Comment (ilteoood) [reply src/upgrade.ts:44]: 4. Use a filter on that array of string, in order to remove this

File: src/upgrade.ts [hunk 6]
```diff
@@ -1,13 +1,69 @@
...
+        return
+      }
+      if (dependencyVersion != latestVersion) {
```
Comment (ilteoood) [inline src/upgrade.ts:47]: 5. apply another filter, in order to also exclude this

File: src/upgrade.ts [hunk 7]
```diff
@@ -1,13 +1,69 @@
...
+        dependencyMap.set(dependencyKey, latestVersion.version)
```
Comment (ilteoood) [inline src/upgrade.ts:48]: 6. use, on the remaining array a for-each in order to update the desired keys

File: src/utils/getLatestPackage.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Returns the Latest version that the package can 
+ * automatically update to following semver.
+ */
+async function getLatestPackage(dependency: Pair): Promise<PubApiModel.Version | undefined> {
+    const dependencyKey: Scalar = dependency.key
+    const dependencyValue: Scalar = dependency.value
+
+    let dependencyName = dependencyKey.value
```
Comment (ilteoood) [inline src/utils/getLatestPackage.ts:14]: why you reassign dependencyKey.value to another variable?
1. Can't you use dependencyValue declared above?
2. Why a "let"? You con't reassign it

File: src/utils/getLatestPackage.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    let dependencyVersion = dependencyValue.value?.replace('^', '') ?? undefined
```
Comment (ilteoood) [inline src/utils/getLatestPackage.ts:15]: you can use a const here

File: src/utils/getLatestPackage.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+    const packageEndpoint = `https://pub.dev/api/packages/${dependencyName}`
+
+    const response = await axios.get(packageEndpoint)
+    const responseData: PubApiModel.RootObject = response.data
+    let versions = responseData.versions
+    let latestVersion: PubApiModel.Version | undefined = undefined
+    versions.reverse().every((element, index) => {
+        let diff = semverDiff(dependencyVersion, element.version)
```
Comment (ilteoood) [inline src/utils/getLatestPackage.ts:24]: const diff = semverDiff(dependencyVersion, element.version);
const isDiffPatch = diff == 'patch';
latestVersion = isDiffPatch ? element : latestVersion;
return !isDiffPatch;

File: src/utils/getLatestPackage.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        if (diff == 'patch') {
```
Comment (ilteoood) [inline src/utils/getLatestPackage.ts:25]: or, in function way, you can (I've not tested this):
const latestVersions = versions.reverse()
.map(element => {element, diff: semverdiff(dependencyVersion, element.version})
.filter(diff => diff.diff == 'patch')
const latestVersion = latestVersions.length > 0 ? latestVersions.pop().element : undefined

Comment (ilteoood) [reply src/utils/getLatestPackage.ts:25]: with this one you can remove the lines from 23 to 30

**Review by ilteoood — COMMENTED** [review]
First reviews

**Comment by preetjdp** [comment]
@ilteoood 
@tianhaoz95 This PR seems unnecessary given `pub outdated / pub upgrade` seems to be fixed with Flutter 1.17
Let me know what you think and lets close this PR

**Comment by tianhaoz95** [comment]
I think a fix on their end is more sustainable for sure. Let me close this one. Feel free to open a new one if you think this PR is partially needed.

**Pull request closed without merging**